### ISSUE TITLE
`ocamlbuild -where` returns a sandbox path when ocamlbuild is built by dune

### ISSUE THREAD

**Issue by gridbugs** [ocaml/dune#10290]
E.g.
```
https://github.com/gridbugs/dune/tree/pkg-hacks?tab=readme-ov-file#package-management-hacks
```

**Comment by gridbugs** [comment]
This feels like a similar issue to https://github.com/ocaml/dune/issues/10291 in that it involves a generated file containing paths to files within the build sandbox.

**Comment by rgrinberg** [comment]
That's right. We should forbid tools from recording any absolute paths. Or at least allow us a way to override them with an environment variable.

**Comment by gridbugs** [comment]
I can't think of a way to fix this that doesn't involve modifying `ocamlbuild`. Any package that persists an opam variable containing a path during its build or install step will be broken in dune's sandbox as all such paths are translated into their corresponding sandboxed path. Fortunately this only seems to affect the ocamlbuild package.

My plan is to modify ocamlbuild so that it only persists paths relative to`OPAM_SWITCH_PREFIX` rather than absolute paths.

**Comment by gridbugs** [comment]
Digging deeper, I'm hesitant to modify ocamlbuild to solve this problem. Ocamlbuild already has compile-time configuration for relocating its installation to a specific destination known at compile-time. The build commands in its opam package are:
```
build: [
  [
    make
    "-f"
    "configure.make"
    "all"
    "OCAMLBUILD_PREFIX=%{prefix}%"
    "OCAMLBUILD_BINDIR=%{bin}%"
    "OCAMLBUILD_LIBDIR=%{lib}%"
    "OCAMLBUILD_MANDIR=%{man}%"
    "OCAML_NATIVE=%{ocaml:native}%"
    "OCAML_NATIVE_TOOLS=%{ocaml:native}%"
  ]
  [make "check-if-preinstalled" "all" "opam-install"]
]
```
This was written under the assumption that `prefix`, `bin`, `lib`, etc are paths within the destination directory structure where ocamlbuild will be installed, which is a reasonable assumption as that is how opam behaves. The only reason this is a problem for dune is it sets those variables to be within the build sandbox. I'm starting to think that we should rethink the decision to translate variables into the sandbox as it deviates from opam's behaviour and will likely lead to more packages that work in opam but not in dune. When a new package is published on opam, it's unreasonable for us to expect that it be tested in dune's build sandbox.

In the short term I propose using an package overlay for ocamlbuild that preprocesses the generated `ocamlbuild_config.ml` file to translate paths out of the sandbox. Short-term we already need to use a modified version of the ocamlbuild package because it contains directory symlinks which dune cannot handle due to a recently-introduced bug (https://github.com/ocaml/dune/issues/9873).

Long term I want to explore alternative ways to effect sandboxing that is consistent with the way that opam builds packages.

See also: https://github.com/ocaml/dune/issues/10316

**Comment by gridbugs** [comment]
We now use a [package overlay](https://github.com/ocaml-dune/opam-overlays/tree/main/packages/ocamlbuild) to install a patched version of ocamlbuild which works around this issue. I have a [PR](https://github.com/ocaml/ocamlbuild/pull/327) to upstream the workaround however since the package overlay exists upstreaming the change isn't strictly necessary. Realistically we aren't going to change the way dune's sandbox works to improve compatibility with opam packages. Only a small number of packages are affected and we've made overlays for the important ones (ocamlfind and ocamlbuild). Closing this since we have implemented a satisfactory solution to the problem and no further work is planned.
